framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,670 @@
+     * example:
+     *
+     * <pre>
+     * <code>
+     *     {@literal @}IntRange(from = 0, to = 10) int a;
+     *     {@literal @}IntRange(from = 3, to = 15) int b;
+     *     ...
+     *     if (a == b) {
+     *         // range of <i>a</i> is now refined to [3, 10] because a value in range [0, 2]
+     *         // cannot be equal to variable <i>b</i> with range [3, 15].
+     *         ...
+     *     }
+     * </code>
+     * </pre>
+     *
+     * @param right the specified {@code Range} to compare with
+     * @return the refined {@code Range}
+     */
+    public Range refineEqualTo(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        long resultFrom = Math.max(from, right.from);
+        long resultTo = Math.min(to, right.to);
+        return createRangeOrNothing(resultFrom, resultTo);
+    }
+
+    /**
+     * Returns the number of possible values enclosed by this range. To prevent overflow, we use
+     * BigInteger for calculation and return a BigInteger.
+     *
+     * @return the number of possible values enclosed by this range
+     */
+    public BigInteger numberOfPossibleValues() {
+        return BigInteger.valueOf(to).subtract(BigInteger.valueOf(from)).add(BigInteger.valueOf(1));
+    }
+
+    /**
+     * Determines if the range is wider than a given value, i.e., if the number of possible values
+     * enclosed by this range is more than the given value.
+     *
+     * @param value the value to compare with
+     * @return true if wider than the given value
+     */
+    public boolean isWiderThan(long value) {
+        return numberOfPossibleValues().compareTo(BigInteger.valueOf(value)) == 1;
+    }
+
+    /**
+     * Determines if this range is completely contained in the scope of the Integer type.
+     *
+     * @return true if the range is completely contained in the scope of the Integer type
+     */
+    public boolean isWithinInteger() {
+        return from >= Integer.MIN_VALUE && to <= Integer.MAX_VALUE;
+    }
+
+    private static final BigInteger longPossibleValues =
+            BigInteger.valueOf(Long.MAX_VALUE - Long.MIN_VALUE + 1);
+
+    /**
+     * Converts a range with BigInteger type bounds to a range with Long type bounds.
+     *
+     * <p>If the BigInteger bounds are out of the Long type scope, convert the bounds to Long type
+     * in accordance with Java overflow rules, e.g., Long.MAX_VALUE + 1 is converted to
+     * Long.MIN_VALUE.
+     *
+     * <p>If the BigInteger range is too wide, i.e., wider than the full range of the Long class,
+     * return EVERYTHING.
+     *
+     * @param bigFrom the lower bound of the BigInteger range
+     * @param bigTo the upper bound of the BigInteger range
+     * @return a range with Long type bounds converted from the BigInteger range
+     */
+    private static Range bigRange2LongRange(BigInteger bigFrom, BigInteger bigTo) {
+        BigInteger numValues = bigTo.subtract(bigFrom);
```

Comment:
This computation looks suspicious.  Should it add 1 as is done for longPossibleValues?  If not, please add a comment explaining why.